private/buf/buflsp/file.go

```diff
@@ -407,132 +542,115 @@ func (f *file) IndexImports(ctx context.Context) {
-		queue = queue[:len(queue)-1]
-
-		if seen[descriptor.Path()] {
-			continue
+	return func(path string) (io.ReadCloser, error) {
+		var uri protocol.URI
+		fileInfo, ok := f.importablePathToObject[path]
+		if ok {
+			uri = protocol.URI("file://" + fileInfo.LocalPath())
+		} else {
+			uri = protocol.URI("file://" + path)
 		}
-		seen[descriptor.Path()] = true
 
-		unused, ok := report.pathToUnusedImports[descriptor.Path()]
-		var unusedIndices []int32
-		if ok {
-			unusedIndices = make([]int32, 0, len(unused))
+		if file := f.Manager().Get(uri); file != nil {
+			return ioext.CompositeReadCloser(strings.NewReader(file.text), ioext.NopCloser), nil
+		} else if !ok {
+			return nil, os.ErrNotExist
 		}
 
-		imports := descriptor.Imports()
-		for i := 0; i < imports.Len(); i++ {
-			dep := imports.Get(i).FileDescriptor
-			if dep == nil {
-				f.lsp.logger.Warn(fmt.Sprintf("found nil FileDescriptor for import %s", imports.Get(i).Path()))
-				continue
-			}
+		return os.Open(fileInfo.LocalPath())
+	}
+}
 
-			queue = append(queue, dep)
+// newAgainstFileOpener returns a fileOpener for building the --against file
+// for this file. In other words, this pulls files out of the git index, if
+// necessary.
+//
+// May return nil, if there is insufficient information to build an --against
+// file.
+func (f *file) newAgainstFileOpener(ctx context.Context) fileOpener {
+	if !f.IsLocal() || f.importablePathToObject == nil || f.gitRemoteBranch == "" {
+		return nil
+	}
 
-			if unused != nil {
-				if _, ok := unused[dep.Path()]; ok {
-					unusedIndices = append(unusedIndices, int32(i))
-				}
-			}
+	return func(path string) (io.ReadCloser, error) {
+		fileInfo, ok := f.importablePathToObject[path]
+		if !ok {
+			return nil, fmt.Errorf("failed to resolve import: %s", path)
 		}
 
-		descriptorProto := protoutil.ProtoFromFileDescriptor(descriptor)
-		if descriptorProto == nil {
-			err = fmt.Errorf("protoutil.ProtoFromFileDescriptor() returned nil for %q", descriptor.Path())
-			break
+		readAtRef := func(ref string) ([]byte, error) {
+			return git.ReadFileAtRef(ctx, f.lsp.runner, f.lsp.container, fileInfo.LocalPath(), ref)
 		}
 
-		var imageFile bufimage.ImageFile
-		imageFile, err = bufimage.NewImageFile(
-			descriptorProto,
-			nil,
-			uuid.UUID{},
-			"",
-			descriptor.Path(),
-			descriptor.Path() != fileInfo.Path(),
-			report.syntaxMissing[descriptor.Path()],
-			unusedIndices,
+		var (
+			data []byte
+			err  error
 		)
-		if err != nil {
-			break
+		switch f.againstKind {
+		case againstTrunk:
+			if f.gitRemote == "" {
+				data, err = readAtRef(f.gitRemoteBranch)
+			} else {
+				data, err = readAtRef(f.gitRemote + "/" + f.gitRemoteBranch)
+			}
+		case againstHead:
+			data, err = readAtRef("HEAD")
 		}
 
-		imageFiles = append(imageFiles, imageFile)
-		f.lsp.logger.Debug(fmt.Sprintf("added image file for %s", descriptor.Path()))
+		if data == nil || errors.Is(err, git.ErrNotARepo) {
+			return os.Open(fileInfo.LocalPath())
```

Comment:
For this case, should we check `f.IsLocal()`?

Reply:
No: the local path is the path on disk, which is the file we need to open for e.g. things that are in cache.